core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+        for {
+          roles <- DBIO
+            .from(samDAO.listUserRolesForResource(SamResourceTypeNames.workspace, workspace.workspaceId, ctx))
+          petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, workspace.googleProjectId))
+          userAndPet = UserAndPetEmails(WorkbenchEmail(ctx.userInfo.userEmail.value), petEmail)
+          _ <- setupProjectRoles(workspace, roles, userAndPet, expirationDate)
+          _ <- setupBucketRoles(workspace, roles, userAndPet, expirationDate)
+          _ <- DBIO
+            .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
+        } yield ()
+      } else {
+        logger.info(
+          s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in workspace ${workspace.toWorkspaceName}"
+        )
+        DBIO.successful()
+      }
+    }
+  }
+
+  def setupFastPassForUserInClonedWorkspace(parentWorkspace: Workspace,
+                                            childWorkspace: Workspace
+  ): ReadWriteAction[Unit] = {
+    if (!config.enabled) {
+      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${parentWorkspace.toWorkspaceName}")
+      return DBIO.successful()
+    }
+
+    DBIO.from(quotaAvailableForClonedWorkspaceFastPassGrants(parentWorkspace, childWorkspace)).flatMap {
+      quotaAvailable =>
+        if (quotaAvailable) {
+          logger.info(
+            s"Adding FastPass access for ${ctx.userInfo.userEmail} in workspace being cloned ${parentWorkspace.toWorkspaceName}"
+          )
+          val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
+          for {
+            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, childWorkspace.googleProjectId))
+            userAndPet = UserAndPetEmails(WorkbenchEmail(ctx.userInfo.userEmail.value), petEmail)
+            _ <- setupBucketRoles(parentWorkspace, Set(SamWorkspaceRoles.reader), userAndPet, expirationDate)
+            _ <- DBIO
+              .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
+          } yield ()
+        } else {
+          logger.info(
+            s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in parent workspace ${parentWorkspace.toWorkspaceName}"
+          )
+          DBIO.successful()
+        }
+    }
+  }
+
+  def removePolicyFastPassesForUser(workspace: Workspace,
+                                    policyName: SamResourcePolicyName,
+                                    email: String
+  ): ReadWriteAction[Unit] =
+    for {
+      maybeUserSubjectId <- DBIO.from(samDAO.getUserIdInfo(email, ctx).map {
+        case SamDAO.User(x) => x.googleSubjectId
+        case _              => None
+      })
+      if maybeUserSubjectId.isDefined
+      userSubjectId = maybeUserSubjectId.orNull
```

Comment:
It looks really odd to me to not have the block starting at line 179 indented, but perhaps there's something different about "if" in Scala?

Reply:
added a clarifying comment, this is actually to short circuit the for-comp (i didnt know you could do that either)

Reply:
I think I understand this now (and the code had changed!).

Reply:
no it is just a line by itself in the for comp that will short circuit the for comp if it is false. We updated with a comment and new lines to make it more clear